File: src/joinagg/joinagg.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type user struct {
	id     string
	region string
	tier   string
}

type event struct {
	id        string
	latencyMS int64
	bytes     int64
}

type joined struct {
	region    string
	tier      string
	latencyMS int64
	bytes     int64
}

func parseUser(line string) (user, bool) {
	first := strings.IndexByte(line, ',')
	second := strings.IndexByte(line[first+1:], ',')
	third := strings.IndexByte(line[first+1+second+1:], ',')
	status := line[first+1+second+1+third+1:]
	if status != "active" {
		return user{}, false
	}
	return user{
		id:     line[:first],
		region: line[first+1 : first+1+second],
		tier:   line[first+1+second+1 : first+1+second+1+third],
	}, true
}

func parseEvent(line string) (event, bool) {
	first := strings.IndexByte(line, ',')
	second := strings.IndexByte(line[first+1:], ',')
	third := strings.IndexByte(line[first+1+second+1:], ',')
	kind := line[first+1 : first+1+second]
	if kind == "noop" {
		return event{}, false
	}
	latencyMS, _ := strconv.ParseInt(line[first+1+second+1:first+1+second+1+third], 10, 64)
	bytes, _ := strconv.ParseInt(line[first+1+second+1+third+1:], 10, 64)
	return event{id: line[:first], latencyMS: latencyMS, bytes: bytes}, true
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Buffer(make([]byte, 1024), 4*1024*1024)
	users := make([]user, 0, 1<<15)
	events := make([]event, 0, 1<<18)
	section := ""
	headerPending := false

	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}
		if line == "[users]" || line == "[events]" {
			section = line
			headerPending = true
			continue
		}
		if headerPending {
			headerPending = false
			continue
		}
		if section == "[users]" {
			if row, ok := parseUser(line); ok {
				users = append(users, row)
			}
		} else if section == "[events]" {
			if row, ok := parseEvent(line); ok {
				events = append(events, row)
			}
		}
	}

	sort.Slice(users, func(i, j int) bool { return users[i].id < users[j].id })
	sort.Slice(events, func(i, j int) bool { return events[i].id < events[j].id })

	joinedRows := make([]joined, 0, len(events))
	for i, j := 0, 0; i < len(users) && j < len(events); {
		if users[i].id < events[j].id {
			i++
		} else if users[i].id > events[j].id {
			j++
		} else {
			current := users[i]
			for j < len(events) && events[j].id == current.id {
				joinedRows = append(joinedRows, joined{
					region:    current.region,
					tier:      current.tier,
					latencyMS: events[j].latencyMS,
					bytes:     events[j].bytes,
				})
				j++
			}
			i++
		}
	}

	sort.Slice(joinedRows, func(i, j int) bool {
		if joinedRows[i].region != joinedRows[j].region {
			return joinedRows[i].region < joinedRows[j].region
		}
		return joinedRows[i].tier < joinedRows[j].tier
	})

	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	for i := 0; i < len(joinedRows); {
		row := joinedRows[i]
		count := int64(0)
		latencySum := int64(0)
		bytesSum := int64(0)
		for i < len(joinedRows) && joinedRows[i].region == row.region && joinedRows[i].tier == row.tier {
			count++
			latencySum += joinedRows[i].latencyMS
			bytesSum += joinedRows[i].bytes
			i++
		}
		fmt.Fprintf(writer, "%s,%s,%d,%d,%d\n", row.region, row.tier, count, latencySum, bytesSum)
	}
	writer.Flush()
}

File: src/joinagg/joinagg.ml
type user = {
  id : string;
  region : string;
  tier : string;
}

type event = {
  id : string;
  latency_ms : int;
  bytes : int;
}

type joined = {
  region : string;
  tier : string;
  latency_ms : int;
  bytes : int;
}

let split_four line =
  let first = String.index line ',' in
  let second = String.index_from line (first + 1) ',' in
  let third = String.index_from line (second + 1) ',' in
  ( String.sub line 0 first,
    String.sub line (first + 1) (second - first - 1),
    String.sub line (second + 1) (third - second - 1),
    String.sub line (third + 1) (String.length line - third - 1) )

let () =
  let users = ref [] in
  let events = ref [] in
  let section = ref "" in
  let header_pending = ref false in
  (try
     while true do
       let line = input_line stdin in
       if line <> "" then begin
         if line = "[users]" || line = "[events]" then begin
           section := line;
           header_pending := true
         end else if !header_pending then
           header_pending := false
         else if !section = "[users]" then begin
           let id, region, tier, status = split_four line in
           if status = "active" then
             users := { id; region; tier } :: !users
         end else if !section = "[events]" then begin
           let id, kind, latency_ms, bytes = split_four line in
           if kind <> "noop" then
             events :=
               { id; latency_ms = int_of_string latency_ms; bytes = int_of_string bytes } :: !events
         end
       end
     done
   with End_of_file -> ());
  let users = Array.of_list !users in
  let events = Array.of_list !events in
  Array.sort (fun (a : user) (b : user) -> String.compare a.id b.id) users;
  Array.sort (fun (a : event) (b : event) -> String.compare a.id b.id) events;
  let joined = ref [] in
  let i = ref 0 in
  let j = ref 0 in
  while !i < Array.length users && !j < Array.length events do
    let user = users.(!i) in
    let event = events.(!j) in
    let cmp = String.compare user.id event.id in
    if cmp < 0 then
      incr i
    else if cmp > 0 then
      incr j
    else begin
      while !j < Array.length events && events.(!j).id = user.id do
        joined :=
          {
            region = user.region;
            tier = user.tier;
            latency_ms = events.(!j).latency_ms;
            bytes = events.(!j).bytes;
          }
          :: !joined;
        incr j
      done;
      incr i
    end
  done;
  let joined = Array.of_list !joined in
  Array.sort
    (fun a b ->
      let region_cmp = String.compare a.region b.region in
      if region_cmp <> 0 then region_cmp else String.compare a.tier b.tier)
    joined;
  let current_region = ref "" in
  let current_tier = ref "" in
  let count = ref 0 in
  let latency_sum = ref 0 in
  let bytes_sum = ref 0 in
  Array.iter
    (fun row ->
      if row.region <> !current_region || row.tier <> !current_tier then begin
        if !current_region <> "" then
          Printf.printf "%s,%s,%d,%d,%d\n" !current_region !current_tier !count !latency_sum !bytes_sum;
        current_region := row.region;
        current_tier := row.tier;
        count := 0;
        latency_sum := 0;
        bytes_sum := 0
      end;
      incr count;
      latency_sum := !latency_sum + row.latency_ms;
      bytes_sum := !bytes_sum + row.bytes)
    joined;
  if !current_region <> "" then
    Printf.printf "%s,%s,%d,%d,%d\n" !current_region !current_tier !count !latency_sum !bytes_sum
